etna/datasets/datasets_generation.py

```diff
@@ -186,3 +186,89 @@ def generate_from_patterns_df(
     df["timestamp"] = pd.date_range(start=start_time, freq=freq, periods=periods)
     df = df.melt(id_vars=["timestamp"], value_name="target", var_name="segment")
     return df
+
+
+def generate_hierarchical_df(
+    periods: int,
+    n_segments: List[int],
+    freq: str = "D",
+    start_time: str = "2000-01-01",
+    ar_coef: Optional[list] = None,
+    sigma: float = 1,
+    random_seed: int = 1,
+) -> pd.DataFrame:
+    """Create DataFrame with AR process data and hierarchical structure.
+
+    Parameters
+    ----------
+    periods:
+        number of timestamps
+    n_segments:
+        number of segments on each level.
+    freq:
+        pandas frequency string for :py:func:`pandas.date_range` that is used to generate timestamp
+    start_time:
+        start timestamp
+    ar_coef:
+        AR coefficients
+    sigma:
+        scale of AR noise
+    random_seed:
+        random seed
+
+    Returns
+    -------
+    pandas.DataFrame
+        DataFrame  at the bottom level of the hierarchy
+    """
+    if len(n_segments) == 0:
+        raise ValueError("`n_segments` should contain at least one positive integer!")
+
+    if (np.less_equal(n_segments, 0)).any():
+        raise ValueError("All `n_segments` elements should be positive!")
+
+    if (np.diff(n_segments) < 0).any():
+        raise ValueError("`n_segments` should represent non-decreasing sequence!")
+
+    rnd = RandomState(seed=random_seed)
+
+    bottom_df = generate_ar_df(
+        periods=periods,
+        start_time=start_time,
+        ar_coef=ar_coef,
+        sigma=sigma,
+        n_segments=n_segments[-1],
+        freq=freq,
+        random_seed=random_seed,
+    )
+
+    bottom_segments = np.unique(bottom_df["segment"])
+
+    child_to_parent = dict()
```

Comment:
Current implementation is pretty straightforward. It randomly assigns each segment at the current level to parent from the next level. It looks a bit bulky because mapping builds on the fly.

Could you elaborate on point 3 of your proposal? In point 2 we are working with the number of level segments and in point 3 we want to index the entire column.